tests/rptest/tests/throughput_limits_snc_test.py

```diff
@@ -204,3 +208,122 @@ def check_node_quota_metric(self, config_prop: self.ConfigProp,
+        """
+        Apply various backpressure to one node of the cluster.
+        Verify that node throughput limits are applied.
+        """
+
+        # add partitions to the topic until there are 3 partition
+        # leaders on any of the nodes
+        partitions_cnt = len(self.redpanda.partitions(self.topic))
+        while True:
+            while True:
+                parts = self.redpanda.partitions(self.topic)
+                if len(parts) < partitions_cnt:
+                    self.logger.debug(
+                        f"New partition not arrived yet, waiting. "
+                        f"{parts} < {partitions_cnt}")
+                    continue
+                parts_c = Counter([p.leader for p in parts])
+                if parts_c[None] == 0:
+                    break
+                self.logger.debug(f"Waiting for leader of the new partition")
+                time.sleep(0.1)
+
+            parts_c_m = parts_c.most_common(1)
+            most_active_node, most_active_count = parts_c_m[0]
+            self.logger.debug(
+                f"Most active node is {most_active_node.name} "
+                f"with {most_active_count} leaders. All stat: {parts_c}")
+            if most_active_count == 3:
+                break
+            assert most_active_count < 3
+
+            self.rpk.add_topic_partitions(self.topic, 1)
+            partitions_cnt += 1
+
+        self.logger.info(f"Working node: {most_active_node.name}")
+        working_partitions = [p for p in parts if p.leader == most_active_node]
+        self.logger.debug(f"Working partitions: {working_partitions}")
+
+        self.redpanda.set_cluster_config({
+            self.ConfigProp.QUOTA_NODE_MAX_IN.value:
+            config_quota_node_max,
+            self.ConfigProp.BAL_PERIOD_MS.value:
+            100,
+        })
+
+        msg_size = 128 * 1024
+        # do as many messages as needed to run for 10s at the set limit
+        # however at smaller limits, more messages are needed to balance
+        # the quota, therefore a constant component is added too
+        msg_count = 100 + int(10 * config_quota_node_max / msg_size)
+        time_to_run = msg_size * msg_count / config_quota_node_max
+        # only test on one of the partitions yet, multi partition tests TBD
+        partition = working_partitions[0].index
+
+        p = RpkProducer(self.test_context,
+                        self.redpanda,
+                        topic=self.topic,
+                        partition=partition,
+                        msg_size=msg_size,
+                        msg_count=msg_count,
+                        produce_timeout=20)
+
+        self.logger.info(
+            f"Starting: msg_size: {msg_size}, msg_count: {msg_count}, time_to_run: {float(f'{time_to_run:.3g}'):g} s, partition: {partition}"
+        )
+        p.start()
+        self.logger.info(f"Started")
+        start = time.time()
+        p.wait(10 + time_to_run)
```

Comment:
Could this lead to 'flaky' tests?

Reply:
The primary reason for flakiness as of now is the precision of time measurement. For example, in the last failed CI run the test spent 2.5 seconds between the test called the producer to start producing, and the moment when the cluster received the 1st produce request. So far no matter what I have tried I could not overcome these gaps. I'm thinking about switching to log analysis for throughput evaluation now.